gum/arch-arm64/gumarm64relocator.c

```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
+              g_hash_table_add (targets_to_check, op->imm);
+            break;
+          }
+          case ARM64_INS_TBZ:
+          case ARM64_INS_TBNZ:
+          {
+            cs_arm64_op * op = &d->operands[2];
+            g_assert (op->type == ARM64_OP_IMM);
+            if (!g_hash_table_contains(checked_targets, op->imm))
+              g_hash_table_add (targets_to_check, op->imm);
+            break;
+          }
+          case ARM64_INS_RET:
+          {
+            go_ahead = false;
+            break;
+          }
+          case ARM64_INS_BR:
+          {
+            go_ahead = false;
+            tragic_exit = true;
+            /*
+             * tragic exit: we stop and reduce to the safest case because
+             * we can not know where the br can jump. Optimization is possible,
+             * by guessing the register value from previouos instruction.
+             */
+            break;
+          }
+          default:
+            break;
+        }
+      }
+      if (tragic_exit)
+        break;
 
-      switch (insn[i].id)
+      g_hash_table_iter_init (&iter, targets_to_check);
+      code_cur = NULL;
+      while (g_hash_table_iter_next (&iter, &key, &key))
       {
-        case ARM64_INS_B:
+        if (g_hash_table_contains (checked_targets, key))
         {
-          cs_arm64_op * op = &d->operands[0];
-          g_assert (op->type == ARM64_OP_IMM);
-          gssize offset =
-              (gssize) op->imm - (gssize) GPOINTER_TO_SIZE (address);
-          if (offset > 0 && offset < (gssize) n)
-            n = offset;
-          eoi = d->cc == ARM64_CC_INVALID || d->cc == ARM64_CC_AL ||
-              d->cc == ARM64_CC_NV;
-          break;
+          g_hash_table_iter_remove (&iter);
+          continue;
         }
-        case ARM64_INS_CBZ:
-        case ARM64_INS_CBNZ:
+        else
         {
-          cs_arm64_op * op = &d->operands[1];
-          g_assert (op->type == ARM64_OP_IMM);
-          gssize offset =
-              (gssize) op->imm - (gssize) GPOINTER_TO_SIZE (address);
-          if (offset > 0 && offset < (gssize) n)
-            n = offset;
+          code_cur = key;
+          if (code_cur > rl.input_cur)
+            address_cur = (code_cur - rl.input_cur) + rl.input_pc;
+          else
+            address_cur = rl.input_pc - (rl.input_cur - code_cur);
+          g_hash_table_iter_remove (&iter);
           break;
         }
-        case ARM64_INS_BR:
-        case ARM64_INS_RET:
-          eoi = TRUE;
-          break;
-        default:
-          break;
       }
+
+      if (code_cur == NULL)
+        break;
     }
 
-    cs_free (insn, count);
+    if (!tragic_exit)
+    {
+      g_hash_table_iter_init (&iter, checked_targets);
+      while (g_hash_table_iter_next (&iter, &key, &key))
+      {
+        gssize offset = (gssize) key - (gssize) GPOINTER_TO_SIZE (address);
+        if (offset > 0 && offset < (gssize) n)
```

Comment:
This came from the old algorithm. My understanding is that we don't care if an instruction jump to the very first instruction, because however we overwrite it, that instruction will jump to the beginning of the trampoline and the behaviour is preserved. Am I wrong?